File: backend/src/main/kotlin/se/daan/moneyprinters/model/game/api/Command.kt
package se.daan.moneyprinters.model.game.api

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo

@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "type"
)
@JsonSubTypes(
        value = [
            JsonSubTypes.Type(value = CreateGame::class, name = "CreateGame"),
            JsonSubTypes.Type(value = AddPlayer::class, name = "AddPlayer"),
            JsonSubTypes.Type(value = StartGame::class, name = "StartGame"),
            JsonSubTypes.Type(value = RollDice::class, name = "RollDice"),
            JsonSubTypes.Type(value = BuyThisSpace::class, name = "BuyThisSpace"),
            JsonSubTypes.Type(value = DeclineThisSpace::class, name = "DeclineThisSpace"),
            JsonSubTypes.Type(value = PlaceBid::class, name = "PlaceBid"),
            JsonSubTypes.Type(value = PassBid::class, name = "PassBid"),
            JsonSubTypes.Type(value = BuyWonBid::class, name = "BuyWonBid"),
            JsonSubTypes.Type(value = DemandRent::class, name = "DemandRent"),
            JsonSubTypes.Type(value = PayRent::class, name = "PayRent"),
            JsonSubTypes.Type(value = EndTurn::class, name = "EndTurn"),
            JsonSubTypes.Type(value = AddOffer::class, name = "AddOffer"),
            JsonSubTypes.Type(value = UpdateOfferValue::class, name = "UpdateOfferValue"),
            JsonSubTypes.Type(value = RemoveOffer::class, name = "RemoveOffer"),
            JsonSubTypes.Type(value = AcceptTrade::class, name = "AcceptTrade"),
            JsonSubTypes.Type(value = RevokeTradeAcceptance::class, name = "RevokeTradeAcceptance"),
        ]
)
sealed class Command

data class CreateGame(
        val gameMaster: String,
        val board: List<Space>,
        val fixedStartMoney: Int,
        val interestRate: Double,
        val returnRate: Double,
) : Command()

data class AddPlayer(
        val id: String,
        val name: String,
        val color: String
) :Command()

data class StartGame(
        val initiator: String
) : Command()

data class RollDice(
        val player: String
) : Command()

data class BuyThisSpace(
        val player: String,
        val cash: Int,
        val borrowed: Int
): Command()

data class DeclineThisSpace(
    val player: String
): Command()

data class PlaceBid(
    val player: String,
    val bid: Int
): Command()

data class PassBid(
    val player: String
): Command()

data class BuyWonBid(
    val player: String,
    val cash: Int,
    val borrowed: Int
): Command()

data class DemandRent(
    val owner: String,
    val demandId: Int
): Command()

data class PayRent(
    val player: String,
    val demandId: Int
): Command()

data class EndTurn(
        val player: String
): Command()

data class AddOffer(
    val from: String,
    val to: String,
    val ownable: String,
    val value: Int
): Command()

data class UpdateOfferValue(
    val from: String,
    val to: String,
    val ownable: String,
    val value: Int
): Command()

data class RemoveOffer(
    val from: String,
    val to: String,
    val ownable: String
): Command()

data class AcceptTrade(
        val by: String,
        val other: String,
        val cashDelta: Int,
        val debtDelta: Int
): Command()

data class RevokeTradeAcceptance(
        val by: String,
        val other: String
): Command()

File: frontend/src/app/game/api/command.ts
export type Command =
  AddPlayer |
  StartGame |
  RollDice |
  BuyThisSpace |
  DeclineThisSpace |
  PlaceBid |
  PassBid |
  BuyWonBid |
  DemandRent |
  PayRent |
  EndTurn |
  AddOffer |
  UpdateOfferValue |
  RemoveOffer |
  AcceptTrade |
  RevokeTradeAcceptance;

export class AddPlayer {
  type: 'AddPlayer' = 'AddPlayer';

  constructor(
    public id: string,
    public name: string,
    public color: string
  ) {
  }
}

export class StartGame {
  type: 'StartGame' = 'StartGame';

  constructor(
    public initiator: string
  ) {
  }
}

export class RollDice {
  type: 'RollDice' = 'RollDice';

  constructor(
    public player: string
  ) {
  }
}

export class BuyThisSpace {
  type:'BuyThisSpace' = 'BuyThisSpace';

  constructor(
    public player: string,
    public cash: number,
    public borrowed: number
  ) {
  }
}

export class DeclineThisSpace {
  type: 'DeclineThisSpace' = 'DeclineThisSpace';

  constructor(
    public player: string
  ) {
  }
}

export class PlaceBid {
  type: 'PlaceBid' = 'PlaceBid';

  constructor(
    public player: string,
    public bid: number
  ) {
  }
}

export class PassBid {
  type: 'PassBid' = 'PassBid';

  constructor(
    public player: string
  ) {
  }
}

export class BuyWonBid {
  type: 'BuyWonBid' = 'BuyWonBid';

  constructor(
    public player: string,
    public cash: number,
    public borrowed: number
  ) {
  }
}


export class DemandRent {
  type: 'DemandRent' = 'DemandRent';

  constructor(
    public owner: string,
    public demandId: number
  ) {
  }
}

export class PayRent {
  type: 'PayRent' = 'PayRent';

  constructor(
    public player: string,
    public demandId: number
  ) {
  }
}

export class EndTurn {
  type: 'EndTurn' = 'EndTurn';

  constructor(
    public player: string
  ) {
  }
}

export class AddOffer {
  type: 'AddOffer' = 'AddOffer';

  constructor(
    public from: string,
    public to: string,
    public ownable: string,
    public value: number
  ) {
  }
}

export class UpdateOfferValue {
  type: 'UpdateOfferValue' = 'UpdateOfferValue';

  constructor(
    public from: string,
    public to: string,
    public ownable: string,
    public value: number
  ) {
  }
}

export class RemoveOffer {
  type: 'RemoveOffer' = 'RemoveOffer';

  constructor(
    public from: string,
    public to: string,
    public ownable: string
  ) {
  }
}

export class AcceptTrade {
  type: 'AcceptTrade' = 'AcceptTrade';

  constructor(
    public from: string,
    public to: string,
    public cashDelta: number,
    public debtDelta: number
  ) {
  }
}

export class RevokeTradeAcceptance {
  type: 'RevokeTradeAcceptance' = 'RevokeTradeAcceptance';

  constructor(
    public from: string,
    public to: string
  ) {
  }
}
